### ISSUE TITLE
Add "rule packs" so people can share their rules easily

### ISSUE THREAD

**Issue by benjaminoakes** [benjaminoakes/maid#50]
Keeping with the Maid theme, we might call them "appliances".

I'd want to be able to have a folder like `~/.maid/appliances` that has a set of appliances.  Rules that would be highly reusable could be stored there, and people could mix and match rules with other users, but not have to get them merged into Maid itself.

For example, this is a highly reusable rule, which might be at `~/.maid/appliances/dump_tmp.rb`:

``` ruby
rule 'Dump temp folder into the trash' do
  trash('~/tmp')
  mkdir('~/tmp')
end
```

Another:

``` ruby
rule 'Move downloaded videos to unwached video folder' do
  to_watch = '~/Videos/To Watch'

  mkdir(to_watch)
  move(dir('~/Downloads/*.webm'), to_watch)
  move(dir('~/Downloads/*.m4v'), to_watch)
  move(dir('~/Downloads/*.mp4'), to_watch)
end
```

**Comment by coaxial** [comment]
This is also an idea I had but I don't know about distribution. Having user create gems for their rules might make the bar too high.

Input welcome 😄

**Comment by coaxial** [comment]
I had an idea on how to do this, inspired by how Golang does third-party dependencies.

Rules could be put in git repos (github and elsewhere), and you'd have a Maid config file where you enumerate the rules you want with the URLs on where to find them. Maid would then go and download a copy of the `rules.rb` at that address and incorporate that into the rules it runs. This sidesteps the whole distribution problem, which also avoids requiring a backend. And it avoids requiring gems to be created, which I think might be too high a burden to users.

There could be a convention to name the repos for these rules `maid-rule-<whatever>` so that they're discoverable when searching by repo name either in a search engine or on git{hub,lab} etc.

Someone could also build a frontend to search for them etc, kinda how https://vimcolorschemes.com/ (and others) show themes but don't curate or host them so maid wouldn't depend on that.

Some unresolved threads with this idea:
- how to separate rules, say I only want this one rule from someone else's `rules.rb`
- convention over configuration would be nice, but I have nothing concrete about this atm
- does made periodically checks if there is a new version of the rule, and does it autoupdate?
- a review system would help avoid dangerous rules since rules get r/w access to pretty much the whole filesystem and mounted (network) drives...

**Comment by benjaminoakes** [comment]
Could definitely work!  How paranoid should we be with respect to abuse and security?

**Comment by coaxial** [comment]
I'd be inclined to do what greasyfork does, put a warning and advise to review the code before using it. I'd want to avoid any kind of moderation or duration because this can quickly become overwhelming and time consuming.

**Comment by benjaminoakes** [comment]
Agreed! But could the code change without a user knowing it? Should it be "pinned" somehow so it can't be changed without review?
